lib/linked_list.rb

```diff
@@ -62,29 +98,71 @@ def length
     # Time Complexity:  
     # Space Complexity
     def get_at_index(index)
-      raise NotImplementedError
+      count = 0
+      current = @head
+      while current
+        if count == index
+          return current.data
+        elsif count > index
+          return nil
+        end
+        current = current.next
+        count += 1
+      end
     end
 
     # method to print all the values in the linked list
     # Time Complexity:  
     # Space Complexity
     def visit
-      raise NotImplementedError
+      values = []
+      current = @head
+      while current != nil
+        values.push(current.data)
+        current = current.next
+      end
+      return values
     end
 
     # method to delete the first node found with specified value
     # Time Complexity:  
     # Space Complexity
     def delete(value)
-      raise NotImplementedError
+      current = @head
+      before = nil
+      while current
+        if current.data == value
+          if !before && !current.next
+            @head = nil
+          elsif !before
+            @head = current.next
+          else
+            before.next = current.next
+          end
+          return "deleted"
+        end
+        before = current
+        current = current.next
+      end
+      return nil
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
     # Time Complexity:  
     # Space Complexity
     def reverse
-      raise NotImplementedError
+      current = @head
+      before = nil
+      after = nil
+
+      while current
+        current.next = before
+        before = current
+        current = current.next

```

Comment:
This shouldn't work, as you've set `current.next` to `nil` right out of the gate.